File: toutiao.py
import requests
import json
from openpyxl import Workbook
import time
import hashlib
import os
import datetime
import pandas as pd

start_url = 'https://www.toutiao.com/api/pc/feed/?category=news_hot&utm_source=toutiao&widen=1&max_behot_time='
url = 'https://www.toutiao.com'

headers = {
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36'
}
cookies = {'tt_webid': '6925668574955062791'}  # 此处cookies可从浏览器中查找，为了避免被头条禁止爬虫

max_behot_time = '0'   # 链接参数
title = []       # 存储新闻标题
source_url = []  # 存储新闻的链接
s_url = []       # 存储新闻的完整链接
source = []      # 存储发布新闻的公众号
media_url = {}   # 存储公众号的完整链接
comments_count = []  # 存储评论量


def get_as_cp():  # 该函数主要是为了获取as和cp参数，程序参考今日头条中的加密js文件：home_4abea46.js
    zz = {}
    now = round(time.time())
    print(now) # 获取当前计算机时间
    e = hex(int(now)).upper()[2:] #hex()转换一个整数对象为16进制的字符串表示 前两位大写。
    print('e:', e)
    a = hashlib.md5()  #hashlib.md5().hexdigest()创建hash对象并返回16进制结果
    print('a:', a)
    a.update(str(int(now)).encode('utf-8'))
    i = a.hexdigest().upper()
    print('i:', i)
    if len(e)!=8:
        zz = {'as':'479BB4B7254C150',
        'cp':'7E0AC8874BB0985'}
        return zz
    n = i[:5]
    a = i[-5:]
    r = ''
    s = ''
    for i in range(5):
        s= s+n[i]+e[i]
    for j in range(5):
        r = r+e[j+3]+a[j]
    zz ={
    'as':'A1'+s+e[-3:],
    'cp':e[0:3]+r+'E1'
    }
    print('zz:', zz)
    return zz

def getdata(url, headers, cookies):  # 解析网页函数
    r = requests.get(url, headers=headers,cookies=cookies)
    print(url)
    data = json.loads(r.text)
    return data


def savedata(title, s_url, source, media_url,comments_count):  # 存储数据到文件
   
    # 存储数据到xlxs文件
    wb = Workbook()
    if not os.path.isdir(os.getcwd()+'/result'):   # 判断文件夹是否存在
        os.makedirs(os.getcwd()+'/result')  # 新建存储文件夹
    filename = os.getcwd()+'/result/result-' + \
        datetime.datetime.now().strftime('%Y-%m-%d-%H-%m')+'.xlsx'  # 新建存储结果的excel文件
    ws = wb.active
    ws.title = 'data'   # 更改工作表的标题
    ws['A1'] = '标题'   # 对表格加入标题
    ws['B1'] = '新闻链接'
    ws['C1'] = '头条号'
    ws['D1'] = '头条号链接'
    ws['E1'] = '评论量'
    for row in range(2, len(title)+2):   # 将数据写入表格
        _ = ws.cell(column=1, row=row, value=title[row-2])
        _ = ws.cell(column=2, row=row, value=s_url[row-2])
        _ = ws.cell(column=3, row=row, value=source[row-2])
        _ = ws.cell(column=4, row=row, value=media_url[source[row-2]])
        _ = ws.cell(column=5, row=row, value=comments_count[row-2])

    wb.save(filename=filename)  # 保存文件
    
    ###按照评论量进行排序并生成排序后的文件
    data = pd.read_excel(filename)
    data.sort_values(by='评论量', ascending=False, inplace=True)
    data.to_csv('result.csv', encoding='utf-8', index=False,
                header=['title', 's_url', 'source', 'media_url', 'comments_count'])
    print(data)



def main(max_behot_time, title, source_url, s_url, source, media_url,comments_count):   # 主函数
    for i in range(10):   # 此处的数字类似于你刷新新闻的次数，正常情况下刷新一次会出现10条新闻，但也存在少于10条的情况；所以最后的结果并不一定是10的倍数
        ascp = get_as_cp()    # 获取as和cp参数的函数
        
        demo = getdata(start_url+max_behot_time+'&max_behot_time_tmp='+max_behot_time +
                       '&tadrequire=true&as='+ascp['as']+'&cp='+ascp['cp'] + '&_signature=', headers, cookies)
        # print(demo)
        time.sleep(1)
        for j in range(len(demo['data'])):
            # print(demo['data'][j]['title'])
            if demo['data'][j]['title'] not in title:
                title.append(demo['data'][j]['title'])  # 获取新闻标题
                source_url.append(demo['data'][j]['source_url'])  # 获取新闻链接
                source.append(demo['data'][j]['source'])  # 获取发布新闻的公众号

            # if demo['data'][j]['comments_count'] not in comments_count:
            if 'comments_count' in demo['data'][j].keys():
                # demo['data'][j]['comments_count'] = demo['data'][j]['comments_count']
                pass
                
            else:
                demo['data'][j]['comments_count'] = 0

            comments_count.append(int(demo['data'][j]['comments_count'])) ##获取评论量

            if demo['data'][j]['source'] not in media_url:
                if 'media_url' in demo['data'][j].keys():
                    media_url[demo['data'][j]['source']] = url + demo['data'][j]['media_url']  # 获取公众号链接
                else:
                    media_url[demo['data'][j]['source']] = ''
            
        print(max_behot_time)

        # 获取下一个链接的max_behot_time参数的值
        max_behot_time = str(demo['next']['max_behot_time'])

        for index in range(len(title)):
            print('标题：', title[index])
            if 'https' not in source_url[index]:
                s_url.append(url+source_url[index])
                print('新闻链接：', url+source_url[index])
            else:
                print('新闻链接：', source_url[index])
                s_url.append(source_url[index])
                # print('源链接：', url+source_url[index])
            print('头条号：', source[index])
            print(len(title))   # 获取的新闻数量


if __name__ == '__main__':
    main(max_behot_time, title, source_url, s_url, source, media_url,comments_count)
    savedata(title, s_url, source, media_url, comments_count)


# https://www.toutiao.com/hot-event/hot-board/?origin=toutiao_pc&_signature=_02B4Z6wo00901eASNswAAIDAHl5K.1czq7HgNjJAAByE29jPOiCW-Hr4LANhznrLGg6.r9kthaUkWpAC8ZjCtQ3SS6Jf6Ag25oW-alkzU6XKceC1XrxI0ACUSOdeCDAHwnJ1pGIxoA-5olqn9e

File: toutiao.js
const CryptoJS = require("crypto-js");
const axios = require("axios");
const puppeteer = require("puppeteer");
const readlineSync = require("readline-sync");

const start_url = "https://m.toutiao.com/list/?tag=__all__&max_time=";
const url = "https://toutiao.com";

const user_url =
  "https://profile.zjurl.cn/api/feed_backflow/profile_share/v1/?app_name=news_article&category=profile_all&stream_api_version=82&request_source=1&appId=1286&appType=mobile_detail_web&isAndroid=true&isIOS=false&isMobile=true&cookie_enabled=true&screen_width=375&screen_height=762&browser_language=zh-CN&browser_platform=MacIntel&browser_name=firefox&browser_version=114.0.0.0&browser_online=true&timezone_name=Asia%2FShanghai";

const headers = {
  "User-Agent":
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36",
  "Accept-Encoding": "gzip, deflate, br",
  "Accept-Language": "zh-CN,zh;q=0.9",
  "Cache-Control:": "no-cache",
  Connection: "keep-alive",
  Host: "m.toutiao.com",
  Pragma: "no-cache",
  "Sec-Ch-Ua":
    '"Not.A/Brand";v="8", "Chromium";v="114", "Google Chrome";v="114"',
  "Sec-Ch-Ua-Mobile": "?0",
  "Sec-Ch-Ua-Platform": "macOS",
  "Sec-Fetch-Dest": "document",
  "Sec-Fetch-Mode": "navigate",
  "Sec-Fetch-Site": "none",
  "Sec-Fetch-User": "?1",
  "Upgrade-Insecure-Requests": "1",
};
const cookies = {
  tt_webid: "7072661175557015053",
};

const puppeteerCookies = {
  value: "7072661175557015053",
  name: "tt_webid",
  domain: "m.toutiao.com",
};

let max_behot_time = "0"; // 链接参数
const source_url = []; // 存储新闻的链接
const s_url = []; // 存储新闻的完整链接
const media_url = []; // 存储公众号的完整链接

const title = []; // 存储新闻标题
const comment_count = []; // 存储评论量
const like_count = []; // 点赞数
const event_url = []; // 新闻链接
const source = []; // 存储发布新闻的公众号
const uid = []; // uid
const items = []; // 聚合数据

let offset = "0";

// const saveDate = () => {};

const getAsCp = () => {
  let result = {};
  let now = Math.floor(Date.now() / 1000);
  // let now = 1686993969;
  console.log("now:", now);
  let now_16 = now.toString(16).toUpperCase();
  console.log("now_16:", now_16);

  let md5 = CryptoJS.MD5(String(now)).toString().toUpperCase();

  console.log("md5:", md5);

  if (now_16.length !== 8) {
    result = { as: "479BB4B7254C150", cp: "7E0AC8874BB0985" };
    return result;
  }

  let first_string = md5.slice(0, 5);
  let second_string = md5.slice(-5);
  let first_result_string = "";
  let second_result_string = "";

  for (let i = 0; i < 5; i++) {
    first_result_string += now_16[i + 3] + second_string[i];
    second_result_string += first_string[i] + now_16[i];
  }

  result = {
    as: "A1" + second_result_string + now_16.slice(-3),
    cp: now_16.slice(0, 3) + first_result_string + "E1",
  };

  console.log("ascp:", result);
  return result;
};

const getData = async (url, headers, cookies) => {
  const resp = await axios.get(url, {
    headers: {
      headers,
      Cookie: cookies,
    },
  });

  console.log("url:", url);

  const data = resp.data;
  return data;
};

const sleep = (time) => {
  return new Promise((res) => {
    setTimeout(() => {
      res();
    }, time);
  });
};

// 接口请求爬取
const main = async (
  max_behot_time,
  title,
  source_url,
  s_url,
  source,
  media_url,
  comment_count
) => {
  let crawlerTimes = readlineSync.question("请输入feed流爬取次数（100）: ");
  if (crawlerTimes == "") crawlerTimes = 100;
  for (let i = 0; i < crawlerTimes; i++) {
    const ascp = getAsCp();

    let res = await getData(
      start_url +
        max_behot_time +
        "&max_behot_time=" +
        max_behot_time +
        "&as=" +
        ascp["as"] +
        "&cp=" +
        ascp["cp"] +
        "&ac=wap&count=20&format=json_raw&aid=1698&_signature=",
      headers,
      cookies
    );

    await sleep(3000);

    for (let j = 0; j < res.data.length; j++) {
      if (!title.find((item) => item === res["data"][j]["title"])) {
        title.push(res["data"][j]["title"]);
        source_url.push(res["data"][j]["source_url"]);
        source.push(res["data"][j]["source"]);
      }
      if (!"comment_count" in Object.keys(res["data"][j])) {
        res["data"][j]["comment_count"] = 0;
      }
      comment_count.push(Number(res["data"][j]["comment_count"]));
      // if (!media_url.find((item) => item === res["data"][j]["source"])) {
      //   if ("media_url" in Object.keys(res["data"][j])) {
      //     media_url[res["data"][j]["source"]] =
      //       url + res["data"][j]["media_url"];
      //   } else {
      //     media_url[res["data"][j]["source"]] = "";
      //   }
      // }
    }

    console.log("max_behot_time:", max_behot_time);
    max_behot_time = String(res.data[res.data.length - 1]["behot_time"]);

    for (let i = 0; i < title.length; i++) {
      console.log("标题：", title[i]);
      if (!source_url[i].startsWith("https")) {
        s_url.push(url + source_url[i]);
        console.log("新闻链接：", url + source_url[i]);
      } else {
        console.lod("新闻链接：", source_url[i]);
        s_url.push(source_url[i]);
      }
      console.log("头条号：", source[i]);
      console.log("评论数：", comment_count[i]);
      console.log(title.length);
    }
  }
};

// main(
//   max_behot_time,
//   title,
//   source_url,
//   s_url,
//   source,
//   media_url,
//   comment_count
// );

// 无头浏览器爬取
const puppeteerAction = async () => {
  let crawlerTimes = readlineSync.question("请输入feed流爬取次数（100）: ");
  if (crawlerTimes == "") crawlerTimes = 100;

  const browser = await puppeteer.launch({
    headless: "new",
    executablePath:
      "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
  });

  const page = await browser.newPage();
  await page.setViewport({
    height: 890,
    width: 375,
    deviceScaleFactor: 1,
    isMobile: true,
    hasTouch: true,
  });

  await page.setCookie(puppeteerCookies);

  for (let i = 0; i < crawlerTimes; i++) {
    const ascp = getAsCp();
    await page.goto(
      start_url +
        max_behot_time +
        "&max_behot_time=" +
        max_behot_time +
        "&as=" +
        ascp["as"] +
        "&cp=" +
        ascp["cp"] +
        "&ac=wap&count=20&format=json_raw&aid=1698&_signature="
    );

    const res = await page.evaluate(async () => {
      return JSON.parse(document.querySelector("pre").innerText);
    });

    await sleep(1000);

    console.log("max_behot_time:", max_behot_time);
    max_behot_time = String(res.data[res.data.length - 1]["behot_time"]);

    for (let j = 0; j < res.data.length; j++) {
      if (!title.find((item) => item === res["data"][j]["title"])) {
        title.push(res["data"][j]["title"]);
        console.log("标题：", res["data"][j]["title"]);

        // source_url.push(res["data"][j]["source_url"]);
        source.push(res["data"][j]["source"]);
        console.log("头条号：", res["data"][j]["source"]);

        event_url.push(res["data"][j]["url"]);
        console.log("新闻链接：", res["data"][j]["url"]);

        if (!"comment_count" in Object.keys(res["data"][j])) {
          comment_count.push(0);
        } else {
          comment_count.push(res["data"][j]["comment_count"]);
        }
        console.log("评论数：", res["data"][j]["comment_count"] || 0);

        if (!"like_count" in Object.keys(res["data"][j])) {
          like_count.push(0);
        } else {
          like_count.push(res["data"][j]["like_count"]);
        }
        console.log("点赞数", res["data"][j]["like_count"] || 0);

        if (!res["data"][j]?.["log_pb"]?.["author_id"]) {
          uid.push("");
        } else {
          uid.push(res["data"][j]?.["log_pb"]?.["author_id"]);
        }
        console.log(
          "作者uid：",
          res["data"][j]?.["log_pb"]?.["author_id"] || ""
        );

        items.push({
          like_count: res["data"][j]["like_count"] || 0,
          uid: res["data"][j]?.["log_pb"]?.["author_id"] || "",
        });
      }
    }

    console.log(title.length);
  }

  items.sort((a, b) => {
    return b.like_count - a.like_count;
  });

  console.log(items);

  await browser.close();
};

const crawlUser = async () => {
  let crawlerTimes = readlineSync.question("请输入大V主页爬取次数（10）: ");
  if (crawlerTimes == "") crawlerTimes = 3;

  const browser = await puppeteer.launch({
    headless: "new",
    executablePath:
      "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
  });

  const page = await browser.newPage();
  await page.setViewport({
    height: 890,
    width: 375,
    deviceScaleFactor: 1,
    isMobile: true,
    hasTouch: true,
  });

  await page.setCookie(puppeteerCookies);

  for (let i = 0; i < crawlerTimes; i++) {
    const url =
      user_url +
      "&visited_uid=" +
      items[0].uid +
      "&user_id=" +
      items[0].uid +
      "&offset=" +
      offset +
      "&_signature=";

    console.log("大V链接：", url);
    await page.goto(url);

    const res = await page.evaluate(async () => {
      return JSON.parse(document.querySelector("pre").innerText);
    });

    await sleep(1000);

    console.log("offset:", offset);
    offset = String(res["offset"]);

    for (let j = 0; j < res.data.length; j++) {
      const result = JSON.parse(res.data[j].content);
      console.log("标题：", result.title);
    }
  }

  await browser.close();
};

(async function () {
  await puppeteerAction();
  await crawlUser();
})();
